### PR TITLE
add notebook example of single image processing

### PR DESCRIPTION
**Pull request by emlynjdavies**
* add a 'first-look' [notebook](https://github.com/SINTEF/pyopia/blob/notebooks/notebooks/single-image-stats.ipynb) to demonstrate processing an image in PyOpia using the pysilcam CNN as an example network
* a couple of bug fixes
* currently limited options to change processing settings in the example in this notebook, as many processing function inputs need updating to get arguments for higher level functions (todo soon in [this issue](https://github.com/SINTEF/pyopia/issues/12))

### CODE REVIEW

**Comment by WillRNaylor** [comment]
The version number will probably go to 8 due to the other PR.

**Comment by davidrwilliamson** [comment]
If it's intended as an example notebook a silc file should be probably included.

**Comment by davidrwilliamson** [comment]
I get an error in cell 3 (#show the image) under Pycharm. I'm using a new conda environment created from environment.yaml. Not sure if the issue is with my setup or the project, usually matplotlib works fine.

ValueError                                Traceback (most recent call last)
File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/IPython/core/formatters.py:339, in __call__(self, obj)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/IPython/core/pylabtools.py:151, in print_figure(fig, fmt, bbox_inches, base64, **kwargs)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/backend_bases.py:2295, in print_figure(self, filename, dpi, facecolor, edgecolor, orientation, format, bbox_inches, pad_inches, bbox_extra_artists, backend, **kwargs)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/artist.py:73, in draw_wrapper(artist, renderer, *args, **kwargs)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/artist.py:50, in draw_wrapper(artist, renderer)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/figure.py:2810, in draw(self, renderer)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/image.py:132, in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/artist.py:50, in draw_wrapper(artist, renderer)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/axes/_base.py:3082, in draw(self, renderer)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/image.py:132, in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/artist.py:50, in draw_wrapper(artist, renderer)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/axis.py:1159, in draw(self, renderer, *args, **kwargs)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/axis.py:1085, in _get_tick_bboxes(self, ticks, renderer)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/axis.py:1085, in <listcomp>(.0)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/text.py:910, in get_window_extent(self, renderer, dpi)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/text.py:309, in _get_layout(self, renderer)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/backends/backend_agg.py:269, in get_text_width_height_descent(self, s, prop, ismath)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/backends/backend_agg.py:303, in _get_agg_font(self, prop)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/font_manager.py:1307, in findfont(self, prop, fontext, directory, fallback_to_default, rebuild_if_missing)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/font_manager.py:1361, in _findfont_cached(self, prop, fontext, directory, fallback_to_default, rebuild_if_missing, rc_params)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/font_manager.py:1307, in findfont(self, prop, fontext, directory, fallback_to_default, rebuild_if_missing)

File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/font_manager.py:1364, in _findfont_cached(self, prop, fontext, directory, fallback_to_default, rebuild_if_missing, rc_params)

ValueError: Failed to find font DejaVu Sans:style=normal:variant=normal:weight=normal:stretch=normal:size=10.0, and fallback to the default font was disabled

**Comment by davidrwilliamson** [comment]
Under Calculate particle statistics, timestamp will fail if you supply a filename that includes a directory path (e.g., my test file of /media/dave/dave_8tb/Mausund/Week_2/Thor/Silcam/D20210420T070622.429715.silc rather than just D20210420T070622.429715.silc). Using os.path.basename() fixes this.

**Comment by davidrwilliamson** [comment]
> > If it's intended as an example notebook a silc file should be probably included.
> 
> yea - i thought about how to do that, but don't want to put image data into the github repo.... the actual example is taken from the pysilcam testdata that is used in pysilcam's github workflows. Any suggestions?

Maybe the test data can be provided by a link like the keras model. Is it already publicly available?

**Comment by emlynjdavies** [comment]
> > > If it's intended as an example notebook a silc file should be probably included.
> > 
> > 
> > yea - i thought about how to do that, but don't want to put image data into the github repo.... the actual example is taken from the pysilcam testdata that is used in pysilcam's github workflows. Any suggestions?
> 
> Maybe the test data can be provided by a link like the keras model. Is it already publicly available?

yea, good idea - I can try to just put a couple of example images on the pysilcam azure blob that could be linked to.

**Comment by emlynjdavies** [comment]
> Under Calculate particle statistics, timestamp will fail if you supply a filename that includes a directory path (e.g., my test file of /media/dave/dave_8tb/Mausund/Week_2/Thor/Silcam/D20210420T070622.429715.silc rather than just D20210420T070622.429715.silc). Using os.path.basename() fixes this.

OK. I have fixed this in [8918b1f](https://github.com/SINTEF/pyopia/pull/10/commits/8918b1f314f21125430379c891772a03a8a73569)

**Comment by emlynjdavies** [comment]
> I get an error in cell 3 (#show the image) under Pycharm. I'm using a new conda environment created from environment.yaml. Not sure if the issue is with my setup or the project, usually matplotlib works fine.
> 
> ValueError Traceback (most recent call last) File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/IPython/core/formatters.py:339, in **call**(self, obj)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/IPython/core/pylabtools.py:151, in print_figure(fig, fmt, bbox_inches, base64, **kwargs)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/backend_bases.py:2295, in print_figure(self, filename, dpi, facecolor, edgecolor, orientation, format, bbox_inches, pad_inches, bbox_extra_artists, backend, **kwargs)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/artist.py:73, in draw_wrapper(artist, renderer, *args, **kwargs)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/artist.py:50, in draw_wrapper(artist, renderer)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/figure.py:2810, in draw(self, renderer)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/image.py:132, in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/artist.py:50, in draw_wrapper(artist, renderer)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/axes/_base.py:3082, in draw(self, renderer)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/image.py:132, in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/artist.py:50, in draw_wrapper(artist, renderer)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/axis.py:1159, in draw(self, renderer, *args, **kwargs)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/axis.py:1085, in _get_tick_bboxes(self, ticks, renderer)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/axis.py:1085, in (.0)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/text.py:910, in get_window_extent(self, renderer, dpi)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/text.py:309, in _get_layout(self, renderer)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/backends/backend_agg.py:269, in get_text_width_height_descent(self, s, prop, ismath)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/backends/backend_agg.py:303, in _get_agg_font(self, prop)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/font_manager.py:1307, in findfont(self, prop, fontext, directory, fallback_to_default, rebuild_if_missing)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/font_manager.py:1361, in _findfont_cached(self, prop, fontext, directory, fallback_to_default, rebuild_if_missing, rc_params)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/font_manager.py:1307, in findfont(self, prop, fontext, directory, fallback_to_default, rebuild_if_missing)
> 
> File ~/anaconda3/envs/pyopia/lib/python3.8/site-packages/matplotlib/font_manager.py:1364, in _findfont_cached(self, prop, fontext, directory, fallback_to_default, rebuild_if_missing, rc_params)
> 
> ValueError: Failed to find font DejaVu Sans:style=normal:variant=normal:weight=normal:stretch=normal:size=10.0, and fallback to the default font was disabled

@davidrwilliamson This looks like a matplotlib / pycharm font rendering issue rather than a pyopia one. Maybe you could try to just load an image and plot it in the PyOpia environment outside of pycharm or in a jupyter noteboook?

File: notebooks/exampledata.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+def get_example_silc_image():
+    url = 'https://pysilcam.blob.core.windows.net/test-data/D20181101T142731.838206.silc'
+    filename = 'D20181101T142731.838206.silc'
+    urllib.request.urlretrieve(url, filename)
+    return filename
+
+
+def get_example_model():
+    # -- Download and unzip the model --#
+    import urllib.request
```
Comment (nepstad) [inline notebooks/exampledata.py:13]: Dulicate import inside function.

File: notebooks/exampledata.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    url = 'https://github.com/SINTEF/PySilCam/wiki/ml_models/keras_model.zip'
+    model_filename = 'keras_model.zip'
+    urllib.request.urlretrieve(url, model_filename)
+    import zipfile
```
Comment (nepstad) [inline notebooks/exampledata.py:17]: Move to top of file?

**Review by nepstad — CHANGES_REQUESTED** [review]

**Review by nepstad — APPROVED** [review]

**Pull request merged**